### ISSUE TITLE
OpenVPN p2p_tls not working

### ISSUE THREAD

**Issue by fenderle** [opnsense/core#2624]
We are running the following versions:

- OPNsense 18.7-amd64
- FreeBSD 11.1-RELEASE-p11
- OpenSSL 1.0.2o 27 Mar 2018

and running into a problem where the configured OpenVPN Peer to Peer (SSL/TLS) Server is not starting due to a configuration file error:

Options error: --client-connect requires --mode server

I tracked down the problem being in the file `/usr/local/etc/inc/plugins.inc.d/openvpn.inc` - if I comment the `case p2p_tls:` from the `switch` block everything works again.It worked in previous versions of OPNsense without any problem.

So this works for me:

`/usr/local/etc/inc/plugins.inc.d/openvpn.inc` around line 630:
```
        // client connect and disconnect handling
        switch ($settings['mode']) {
            case 'server_user':
            case 'server_tls_user':
                $conf .= "client-disconnect \"/usr/local/etc/inc/plugins.inc.d/openvpn/attributes.sh {$mode_id}\"\n";
                break;
            case 'server_tls':
//            case 'p2p_tls':
                // For non user auth types setup client specific overrides,
                // user authenticated ones are commissioned using the auth
                // script in option auth-user-pass-verify.
                $conf .= "client-connect \"/usr/local/etc/inc/plugins.inc.d/openvpn/ovpn_setup_cso.php {$mode_id}\"\n";
                break;
            default:
                break;
        }
```

**Comment by fichtner** [comment]
It was already fixed in 18.7 but it seems there is a strange thing going on a bit later:

https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn.inc#L661-L671

It seems you neither have a server or client directive because of this code. What is your subnet bit mask?


Cheers,
Franco

**Comment by fichtner** [comment]
Relevant 18.7 fix... https://github.com/opnsense/core/commit/24f1d05c7714

**Comment by fenderle** [comment]
It's a /30 - here are my settings (settings with no value or being unticked have been omitted):

- Server Mode: Peer to Peer ( SSL/TLS )
- Protocol: UDP
- Device Mode: tun
- Interface: WAN
- Local Port: 1196
- TLS Authentication: enabled (and TLS Key present)
- Peer Certificate Authority: CA certificate
- Peer Certificate Revocation List: CA's CRL
- Server Certificate: Server certificate
- DH Parameters Length: 1024 bit
- Encryption algorithm: AES-256-CBC
- Auth Digest Algorithm: SHA1
- Hardware Crypto: No Hardware Crypto Acceleration
- Certificate Depth: One (Client+Server)
- IPv4 Tunnel Network: 10.254.253.0/30
- Compression: Enabled with Adaptive Compression
- Verbosity Level: 3

Thats the generated config file (not working):

```
dev ovpns1
verb 3
dev-type tun
tun-ipv6
dev-node /dev/tun1
writepid /var/run/openvpn_server1.pid
#user nobody
#group nobody
script-security 3
daemon
keepalive 10 60
ping-timer-rem
persist-tun
persist-key
proto udp
cipher AES-256-CBC
auth SHA1
up /usr/local/etc/inc/plugins.inc.d/openvpn/ovpn-linkup
down /usr/local/etc/inc/plugins.inc.d/openvpn/ovpn-linkdown
local 192.168.1.2
client-connect "/usr/local/etc/inc/plugins.inc.d/openvpn/ovpn_setup_cso.php server1"
tls-server
ifconfig 10.254.253.1 10.254.253.2
tls-verify "/usr/local/etc/inc/plugins.inc.d/openvpn/ovpn_auth_verify tls 'xxxx_redacted_xxxx' 1"
lport 1196
management /var/etc/openvpn/server1.sock unix
ca /var/etc/openvpn/server1.ca
cert /var/etc/openvpn/server1.cert
key /var/etc/openvpn/server1.key
dh /usr/local/etc/dh-parameters.1024.sample
crl-verify /var/etc/openvpn/server1.crl-verify
tls-auth /var/etc/openvpn/server1.tls-auth 0
comp-lzo adaptive
```

**Comment by AdSchellevis** [comment]
this https://github.com/opnsense/core/commit/29cb71425d0428833d52489b97589f7465a8a164 should fix the /30 issue, although I expect there are some more rough edges here. 

Without the server directive the client-connect script can't be used, which is hard to know because the server parameter isn't stored separately, hence the additional `if` construction in this fix (which aligns to the part found earlier).

**Comment by fichtner** [comment]
Thanks for fixing @AdSchellevis 

@fenderle you can install the patch via:

    # opnsense-patch 29cb714

It won't be present in 18.7.1. Builds are currently finishing up, but you can always reapply after update.

**Comment by fenderle** [comment]
Patch works - that closes the issue for me. Thank you.

**Comment by fichtner** [comment]
Thanks, will be fixed in 18.7.2.
